File: ml/build_corpus.js
const fs = require("fs");
const path = require("path");

const ROOT = __dirname;
const RAW = path.join(ROOT, "raw");
const OUTPUT = path.join(ROOT, "corpus.csv");

const LANGUAGES = {
  english: { manualColumn: "english", floresCode: "eng_Latn", masakhanewsCode: "eng" },
  yoruba: { manualColumn: "yoruba", floresCode: "yor_Latn", masakhanewsCode: "yor" },
  hausa: { manualColumn: "hausa", floresCode: "hau_Latn", masakhanewsCode: "hau" },
  igbo: { manualColumn: "igbo", floresCode: "ibo_Latn", masakhanewsCode: "ibo" },
};

function cleanText(text) {
  return String(text || "").replace(/^\uFEFF/, "").trim().replace(/\s+/g, " ");
}

function parseCsvLine(line, delimiter = ",") {
  const values = [];
  let current = "";
  let quoted = false;

  for (let i = 0; i < line.length; i += 1) {
    const char = line[i];
    const next = line[i + 1];

    if (char === '"' && quoted && next === '"') {
      current += '"';
      i += 1;
    } else if (char === '"') {
      quoted = !quoted;
    } else if (char === delimiter && !quoted) {
      values.push(current);
      current = "";
    } else {
      current += char;
    }
  }

  values.push(current);
  return values;
}

function parseDelimited(content, delimiter = ",") {
  const lines = content.replace(/\r\n/g, "\n").replace(/\r/g, "\n").split("\n").filter(Boolean);
  const headers = parseCsvLine(lines[0], delimiter);
  return lines.slice(1).map((line) => {
    const values = parseCsvLine(line, delimiter);
    return Object.fromEntries(headers.map((header, index) => [header, values[index] || ""]));
  });
}

function csvEscape(value) {
  const text = String(value ?? "");
  if (/[",\n\r]/.test(text)) {
    return `"${text.replace(/"/g, '""')}"`;
  }
  return text;
}

function addRow(rows, text, language, source, domain = "", intent = "") {
  const cleaned = cleanText(text);
  if (cleaned.length < 3) return;
  rows.push({ text: cleaned, language, source, domain, intent });
}

function loadManualHealthcare() {
  const content = fs.readFileSync(path.join(ROOT, "manual_healthcare_samples.csv"), "utf8");
  const sourceRows = parseDelimited(content, ",");
  const rows = [];

  for (const item of sourceRows) {
    for (const [language, config] of Object.entries(LANGUAGES)) {
      addRow(
        rows,
        item[config.manualColumn],
        language,
        "manual_healthcare_google_translate",
        item.domain,
        item.intent,
      );
    }
  }

  return rows;
}

function loadManualStressTraining() {
  const content = fs.readFileSync(path.join(ROOT, "manual_stress_training_samples.csv"), "utf8");
  const sourceRows = parseDelimited(content, ",");
  const rows = [];

  for (const item of sourceRows) {
    addRow(
      rows,
      item.text,
      item.language,
      "manual_stress_training",
      item.domain,
      item.case_type,
    );
  }

  return rows;
}

function loadFlores() {
  const rows = [];
  const base = path.join(RAW, "flores", "flores200_dataset");

  for (const [language, config] of Object.entries(LANGUAGES)) {
    for (const split of ["dev", "devtest"]) {
      const filePath = path.join(base, split, `${config.floresCode}.${split}`);
      const lines = fs.readFileSync(filePath, "utf8").split(/\r?\n/);
      for (const line of lines) {
        addRow(rows, line, language, `flores200_${split}`, "general", "");
      }
    }
  }

  return rows;
}

function loadMasakhanews(maxPerLanguage = 1000) {
  const rows = [];

  for (const [language, config] of Object.entries(LANGUAGES)) {
    const languageRows = [];
    const folder = path.join(RAW, "masakhanews", config.masakhanewsCode);

    for (const split of ["train", "dev", "test"]) {
      const filePath = path.join(folder, `${split}.tsv`);
      const content = fs.readFileSync(filePath, "utf8");
      const sourceRows = parseDelimited(content, "\t");
      for (const item of sourceRows) {
        addRow(
          languageRows,
          `${item.headline || ""} ${item.text || ""}`,
          language,
          `masakhanews_${split}`,
          item.category || "",
          "",
        );
      }
    }

    rows.push(...languageRows.slice(0, maxPerLanguage));
  }

  return rows;
}

function deduplicate(rows) {
  const seen = new Set();
  const unique = [];

  for (const row of rows) {
    const key = `${row.language}\u0000${row.text.toLocaleLowerCase()}`;
    if (seen.has(key)) continue;
    seen.add(key);
    unique.push(row);
  }

  return unique;
}

function balanceRows(rows) {
  const grouped = Object.fromEntries(Object.keys(LANGUAGES).map((language) => [language, []]));
  for (const row of rows) grouped[row.language].push(row);
  const target = Math.min(...Object.values(grouped).map((items) => items.length));
  return Object.values(grouped).flatMap((items) => items.slice(0, target));
}

function writeCorpus(rows) {
  const headers = ["text", "language", "source", "domain", "intent"];
  const lines = [headers.join(",")];
  for (const row of rows) {
    lines.push(headers.map((header) => csvEscape(row[header])).join(","));
  }
  fs.writeFileSync(OUTPUT, `${lines.join("\n")}\n`, "utf8");
}

function main() {
  let rows = [
    ...loadManualHealthcare(),
    ...loadManualStressTraining(),
    ...loadFlores(),
    ...loadMasakhanews(1000),
  ];

  rows = deduplicate(rows);
  rows = balanceRows(rows);
  writeCorpus(rows);

  console.log(`Saved: ${OUTPUT}`);
  console.log(`Total rows: ${rows.length}`);
  for (const language of Object.keys(LANGUAGES)) {
    console.log(`${language}: ${rows.filter((row) => row.language === language).length}`);
  }
}

main();

File: ml/build_corpus.py
import csv
import random
from pathlib import Path


ROOT = Path(__file__).resolve().parent
RAW = ROOT / "raw"
OUTPUT = ROOT / "corpus.csv"

RANDOM_SEED = 42

LANGUAGES = {
    "english": {
        "manual_column": "english",
        "flores_code": "eng_Latn",
        "masakhanews_code": "eng",
    },
    "yoruba": {
        "manual_column": "yoruba",
        "flores_code": "yor_Latn",
        "masakhanews_code": "yor",
    },
    "hausa": {
        "manual_column": "hausa",
        "flores_code": "hau_Latn",
        "masakhanews_code": "hau",
    },
    "igbo": {
        "manual_column": "igbo",
        "flores_code": "ibo_Latn",
        "masakhanews_code": "ibo",
    },
}


def clean_text(text: str) -> str:
    return " ".join(str(text).replace("\ufeff", "").split())


def add_row(rows: list[dict[str, str]], text: str, language: str, source: str, domain: str = "", intent: str = "") -> None:
    cleaned = clean_text(text)
    if len(cleaned) < 3:
        return
    rows.append(
        {
            "text": cleaned,
            "language": language,
            "source": source,
            "domain": domain,
            "intent": intent,
        }
    )


def load_manual_healthcare() -> list[dict[str, str]]:
    rows: list[dict[str, str]] = []
    path = ROOT / "manual_healthcare_samples.csv"

    with path.open("r", encoding="utf-8-sig", newline="") as file:
        reader = csv.DictReader(file)
        for item in reader:
            for language, config in LANGUAGES.items():
                add_row(
                    rows,
                    item[config["manual_column"]],
                    language,
                    "manual_healthcare_google_translate",
                    item["domain"],
                    item["intent"],
                )

    return rows


def load_manual_stress_training() -> list[dict[str, str]]:
    rows: list[dict[str, str]] = []
    path = ROOT / "manual_stress_training_samples.csv"

    with path.open("r", encoding="utf-8-sig", newline="") as file:
        reader = csv.DictReader(file)
        for item in reader:
            add_row(
                rows,
                item["text"],
                item["language"],
                "manual_stress_training",
                item["domain"],
                item["case_type"],
            )

    return rows


def load_flores() -> list[dict[str, str]]:
    rows: list[dict[str, str]] = []
    base = RAW / "flores" / "flores200_dataset"

    for language, config in LANGUAGES.items():
        for split in ("dev", "devtest"):
            path = base / split / f"{config['flores_code']}.{split}"
            with path.open("r", encoding="utf-8") as file:
                for line in file:
                    add_row(rows, line, language, f"flores200_{split}", "general", "")

    return rows


def load_masakhanews(max_per_language: int = 1000) -> list[dict[str, str]]:
    rng = random.Random(RANDOM_SEED)
    rows: list[dict[str, str]] = []

    for language, config in LANGUAGES.items():
        language_rows: list[dict[str, str]] = []
        folder = RAW / "masakhanews" / config["masakhanews_code"]

        for split in ("train", "dev", "test"):
            path = folder / f"{split}.tsv"
            with path.open("r", encoding="utf-8-sig", newline="") as file:
                reader = csv.DictReader(file, delimiter="\t")
                for item in reader:
                    text = f"{item.get('headline', '')} {item.get('text', '')}"
                    add_row(language_rows, text, language, f"masakhanews_{split}", item.get("category", ""), "")

        rng.shuffle(language_rows)
        rows.extend(language_rows[:max_per_language])

    return rows


def deduplicate(rows: list[dict[str, str]]) -> list[dict[str, str]]:
    seen: set[tuple[str, str]] = set()
    unique_rows: list[dict[str, str]] = []

    for row in rows:
        key = (row["language"], row["text"].casefold())
        if key in seen:
            continue
        seen.add(key)
        unique_rows.append(row)

    return unique_rows


def balance_rows(rows: list[dict[str, str]]) -> list[dict[str, str]]:
    rng = random.Random(RANDOM_SEED)
    grouped: dict[str, list[dict[str, str]]] = {language: [] for language in LANGUAGES}

    for row in rows:
        grouped[row["language"]].append(row)

    target = min(len(items) for items in grouped.values())
    balanced: list[dict[str, str]] = []

    for language, items in grouped.items():
        rng.shuffle(items)
        balanced.extend(items[:target])

    rng.shuffle(balanced)
    return balanced


def write_corpus(rows: list[dict[str, str]]) -> None:
    fieldnames = ["text", "language", "source", "domain", "intent"]
    with OUTPUT.open("w", encoding="utf-8", newline="") as file:
        writer = csv.DictWriter(file, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)


def print_summary(rows: list[dict[str, str]]) -> None:
    print(f"Saved: {OUTPUT}")
    print(f"Total rows: {len(rows)}")
    for language in LANGUAGES:
        count = sum(1 for row in rows if row["language"] == language)
        print(f"{language}: {count}")


def main() -> None:
    rows: list[dict[str, str]] = []
    rows.extend(load_manual_healthcare())
    rows.extend(load_manual_stress_training())
    rows.extend(load_flores())
    rows.extend(load_masakhanews(max_per_language=1000))

    rows = deduplicate(rows)
    rows = balance_rows(rows)
    write_corpus(rows)
    print_summary(rows)


if __name__ == "__main__":
    main()
